### ISSUE TITLE
java.util.NoSuchElementException: Key Image Name is missing in the map

### ISSUE THREAD

**Issue by BenjaminBini** [ayedo/tcpkiller#1]
Hi!

I got these errors after installing on Windows 10 :

![image](https://user-images.githubusercontent.com/1637397/77961478-25277000-72da-11ea-9872-7dfae8652df6.png)

![image](https://user-images.githubusercontent.com/1637397/77961501-2f496e80-72da-11ea-8292-c30151e8d559.png)

![image](https://user-images.githubusercontent.com/1637397/77961530-42f4d500-72da-11ea-8ce1-d93558bb19dd.png)

**Comment by ayedo** [comment]
Hello BenjaminBibi, thanks for your report.

Can you please do the following on your machine, and reply here with the output? 

Run the following command in `cmd` or powershell:

`tasklist /svc /fo csv`

Thanks.

**Comment by BenjaminBini** [comment]
Here you go:

```"Nom de l’image","PID","Services"
"System Idle Process","0","N/A"
"System","4","N/A"
"Registry","120","N/A"
"smss.exe","492","N/A"
"csrss.exe","784","N/A"
"wininit.exe","872","N/A"
"csrss.exe","880","N/A"
"services.exe","944","N/A"
"lsass.exe","952","EFS,KeyIso,SamSs,VaultSvc"
"svchost.exe","656","PlugPlay"
"svchost.exe","600","BrokerInfrastructure,DcomLaunch,Power,SystemEventsBroker"
"fontdrvhost.exe","988","N/A"
"svchost.exe","1076","RpcEptMapper,RpcSs"
"WUDFHost.exe","1128","N/A"
"svchost.exe","1196","LSM"
"winlogon.exe","1244","N/A"
"fontdrvhost.exe","1308","N/A"
"dwm.exe","1416","N/A"
"svchost.exe","1496","TermService"
"svchost.exe","1552","BTAGService"
"svchost.exe","1564","BthAvctpSvc"
"svchost.exe","1572","bthserv"
"svchost.exe","1728","NcbService"
"svchost.exe","1736","TimeBrokerSvc"
"svchost.exe","1912","EventLog"
"svchost.exe","1928","hidserv"
"svchost.exe","2012","Schedule"
"svchost.exe","864","ProfSvc"
"svchost.exe","1300","PhoneSvc"
"svchost.exe","1364","SEMgrSvc"
"svchost.exe","2152","nsi"
"svchost.exe","2200","Dhcp"
"svchost.exe","2256","UserManager"
"svchost.exe","2292","CoreMessagingRegistrar"
"NVDisplay.Container.exe","2344","NVDisplay.ContainerLocalSystem"
"svchost.exe","2436","NlaSvc"
"svchost.exe","2484","Dnscache"
"svchost.exe","2588","Themes"
"svchost.exe","2596","SysMain"
"svchost.exe","2604","EventSystem"
"svchost.exe","2712","DeviceAssociationService"
"svchost.exe","2760","netprofm"
"svchost.exe","2804","SENS"
"Memory Compression","2832","N/A"
"svchost.exe","2884","FontCache"
"svchost.exe","2892","AudioEndpointBuilder"
"dasHost.exe","2920","N/A"
"svchost.exe","2052","UmRdpService"
"svchost.exe","884","DispBrokerDesktopSvc"
"svchost.exe","3160","WinHttpAutoProxySvc"
"NVDisplay.Container.exe","3240","N/A"
"svchost.exe","3300","CertPropSvc"
"svchost.exe","3360","LanmanWorkstation"
"svchost.exe","3440","SessionEnv"
"svchost.exe","3640","SSDPSRV"
"svchost.exe","3680","Audiosrv"
"svchost.exe","3852","Wcmsvc"
"svchost.exe","3860","DusmSvc"
"svchost.exe","4160","WlanSvc"
"svchost.exe","4248","ShellHWDetection"
"svchost.exe","4296","BFE,mpssvc"
"svchost.exe","4304","StateRepository"
"scheduler.exe","4468","FA_Scheduler"
"svchost.exe","4516","Winmgmt"
"svchost.exe","4616","iphlpsvc"
"spoolsv.exe","4908","Spooler"
"svchost.exe","5028","AzureAttestService"
"svchost.exe","5060","CryptSvc"
"svchost.exe","5092","DiagTrack"
"com.docker.service","5100","com.docker.service"
"svchost.exe","5108","DPS"
"AdobeUpdateService.exe","3780","AdobeUpdateService"
"OfficeClickToRun.exe","5132","ClickToRunSvc"
"IpOverUsbSvc.exe","5144","IpOverUsbSvc"
"AGSService.exe","5164","AGSService"
"AGMService.exe","5184","AGMService"
"nvcontainer.exe","5208","NvContainerLocalSystem"
"svchost.exe","5236","SstpSvc"
"svchost.exe","5248","stisvc"
"svchost.exe","5288","TrkWks"
"svchost.exe","5312","WpnService"
"RtkAudUService64.exe","5320","RtkAudioUniversalService"
"vmware-authd.exe","5340","VMAuthdService"
"pg_ctl.exe","5368","postgresql-x64-9.3"
"dvt-jb_licsrv.386.exe","5396","JetBrainsLicServerDVT"
"vmware-usbarbitrator64.exe","5404","VMUSBArbService"
"mysqld.exe","5412","MySQL56"
"svchost.exe","5428","TapiSrv"
"vmnetdhcp.exe","5452","VMnetDHCP"
"ss_conn_service.exe","5460","ss_conn_service"
"vmnat.exe","5472","VMware NAT Service"
"MsMpEng.exe","5512","WinDefend"
"OriginWebHelperService.exe","5604","Origin Web Helper Service"
"mysqld.exe","5732","MySQL80"
"vmms.exe","5780","vmms"
"svchost.exe","5808","WdiServiceHost"
"svchost.exe","5920","LanmanServer"
"TeamViewer_Service.exe","6016","TeamViewer"
"nfsclnt.exe","6032","NfsClnt"
"svchost.exe","6108","RasMan"
"postgres.exe","6816","N/A"
"conhost.exe","6824","N/A"
"sqlservr.exe","6904","MSSQLSERVER"
"sqlservr.exe","6928","MSSQL$SQLEXPRESS"
"mysqld.exe","7080","N/A"
"conhost.exe","7088","N/A"
"WmiPrvSE.exe","7248","N/A"
"unsecapp.exe","7292","N/A"
"postgres.exe","7448","N/A"
"svchost.exe","7488","PolicyAgent"
"postgres.exe","7588","N/A"
"postgres.exe","7596","N/A"
"postgres.exe","7604","N/A"
"postgres.exe","7612","N/A"
"postgres.exe","7620","N/A"
"rundll32.exe","7552","N/A"
"vmcompute.exe","8524","vmcompute"
"svchost.exe","8768","fdPHost"
"svchost.exe","8860","FDResPub"
"svchost.exe","9176","hns"
"svchost.exe","8488","nvagent"
"Launchpad.exe","10220","MSSQLLaunchpad$SQLEXPRESS"
"fdlauncher.exe","9500","MSSQLFDLauncher$SQLEXPRESS"
"FCDBLog.exe","9384","N/A"
"fdhost.exe","9840","N/A"
"conhost.exe","9832","N/A"
"SearchIndexer.exe","10580","WSearch"
"FortiESNAC.exe","10876","N/A"
"FortiSSLVPNdaemon.exe","10904","N/A"
"FortiSettings.exe","10920","N/A"
"dllhost.exe","11244","N/A"
"nvcontainer.exe","11344","N/A"
"sihost.exe","11364","N/A"
"nvcontainer.exe","11372","N/A"
"svchost.exe","11436","CDPUserSvc_c3ccd"
"svchost.exe","11556","LxssManagerUser_c3ccd"
"svchost.exe","11592","WpnUserService_c3ccd"
"svchost.exe","11712","TokenBroker"
"svchost.exe","11876","TabletInputService"
"ctfmon.exe","11928","N/A"
"taskhostw.exe","12152","N/A"
"explorer.exe","12244","N/A"
"NisSrv.exe","4332","WdNisSvc"
"svchost.exe","992","CDPSvc"
"svchost.exe","11032","cbdhsvc_c3ccd"
"svchost.exe","12604","OneSyncSvc_c3ccd,PimIndexMaintenanceSvc_c3ccd,UnistoreSvc_c3ccd,UserDataSvc_c3ccd"
"svchost.exe","12780","NgcSvc"
"svchost.exe","12836","NgcCtnrSvc"
"FortiTray.exe","12872","N/A"
"svchost.exe","12964","SharedAccess"
"StartMenuExperienceHost.exe","12688","N/A"
"RuntimeBroker.exe","13504","N/A"
"RuntimeBroker.exe","14148","N/A"
"svchost.exe","13128","LicenseManager"
"SkypeBackgroundHost.exe","11832","N/A"
"svchost.exe","13428","InstallService"
"SkypeApp.exe","14108","N/A"
"NVIDIA Web Helper.exe","9276","N/A"
"conhost.exe","11772","N/A"
"GoogleCrashHandler.exe","1444","N/A"
"GoogleCrashHandler64.exe","13652","N/A"
"RuntimeBroker.exe","12712","N/A"
"RuntimeBroker.exe","14532","N/A"
"svchost.exe","14832","Appinfo"
"SkypeBridge.exe","14876","N/A"
"dllhost.exe","15008","N/A"
"RuntimeBroker.exe","15192","N/A"
"svchost.exe","15720","PcaSvc"
"nvsphelper64.exe","13448","N/A"
"NVIDIA Share.exe","13368","N/A"
"RtkAudUService64.exe","15888","N/A"
"NVIDIA Share.exe","16184","N/A"
"NVIDIA Share.exe","16268","N/A"
"LogiOptions.exe","14804","N/A"
"LogiOptionsMgr.exe","16556","N/A"
"LogiOverlay.exe","16580","N/A"
"googledrivesync.exe","16768","N/A"
"googledrivesync.exe","14588","N/A"
"CCleaner64.exe","16944","N/A"
"OneDrive.exe","9184","N/A"
"AppleMobileDeviceProcess.exe","18504","N/A"
"OUTLOOK.EXE","18696","N/A"
"Spotify.exe","19228","N/A"
"ONENOTEM.EXE","19408","N/A"
"Spotify.exe","19544","N/A"
"Spotify.exe","19824","N/A"
"Spotify.exe","20348","N/A"
"Spotify.exe","20404","N/A"
"svchost.exe","20464","WebClient"
"FileCoAuth.exe","21036","N/A"
"svchost.exe","22116","wscsvc"
"SecurityHealthService.exe","22296","SecurityHealthService"
"ShellExperienceHost.exe","17516","N/A"
"RuntimeBroker.exe","17688","N/A"
"elasticsearch-service-x64.exe","17356","elasticsearch-service-x64"
"conhost.exe","23428","N/A"
"jhi_service.exe","18852","jhi_service"
"LMS.exe","18588","LMS"
"SgrmBroker.exe","18908","SgrmBroker"
"svchost.exe","21132","UsoSvc"
"controller.exe","18680","N/A"
"conhost.exe","11132","N/A"
"ApplicationFrameHost.exe","15216","N/A"
"WinStore.App.exe","22452","N/A"
"RuntimeBroker.exe","2276","N/A"
"AdobeNotificationClient.exe","24176","N/A"
"SystemSettings.exe","22316","N/A"
"RuntimeBroker.exe","16492","N/A"
"svchost.exe","18768","WbioSrvc"
"WindowsInternal.ComposableShell.Experiences.TextInput.InputApp.exe","18232","N/A"
"WINWORD.EXE","8584","N/A"
"FoxitProxyServer_Socket_RD.exe","22952","N/A"
"FoxitProxyServer_Socket_PH.exe","23028","N/A"
"svchost.exe","11208","PrintWorkflowUserSvc_c3ccd"
"svchost.exe","12788","StorSvc"
"svchost.exe","23512","DoSvc"
"Unsplashed.exe","1644","N/A"
"isa.exe","11696","Intel(R) Security Assist"
"Video.UI.exe","18444","N/A"
"RuntimeBroker.exe","13580","N/A"
"svchost.exe","12144","Netman"
"SettingSyncHost.exe","3740","N/A"
"RuntimeBroker.exe","2092","N/A"
"svchost.exe","12808","LxssManager"
"dllhost.exe","23756","N/A"
"Microsoft.Photos.exe","21512","N/A"
"svchost.exe","8676","lfsvc"
"svchost.exe","15252","DsSvc"
"RuntimeBroker.exe","17996","N/A"
"Code.exe","25772","N/A"
"Code.exe","26396","N/A"
"Code.exe","21336","N/A"
"Code.exe","15032","N/A"
"Code.exe","26352","N/A"
"Code.exe","21124","N/A"
"Code.exe","26348","N/A"
"svchost.exe","24428","AppMgmt"
"YourPhoneServer.exe","16804","N/A"
"svchost.exe","8100","QWAVE"
"WINWORD.EXE","22024","N/A"
"ONENOTE.EXE","21276","N/A"
"taskhostw.exe","18392","N/A"
"lghub_updater.exe","19184","LGHUBUpdaterService"
"lghub.exe","10764","N/A"
"lghub_agent.exe","2664","N/A"
"lghub.exe","30216","N/A"
"unsecapp.exe","29292","N/A"
"CompPkgSrv.exe","19152","N/A"
"rundll32.exe","21784","N/A"
"TGitCache.exe","20876","N/A"
"mRemoteNG.exe","96","N/A"
"WmiPrvSE.exe","2928","N/A"
"pageant.exe","22880","N/A"
"PuTTYNG.exe","5972","N/A"
"idea64.exe","23716","N/A"
"fsnotifier64.exe","14764","N/A"
"conhost.exe","16000","N/A"
"firefox.exe","30680","N/A"
"firefox.exe","27872","N/A"
"firefox.exe","29988","N/A"
"Taskmgr.exe","24224","N/A"
"firefox.exe","26320","N/A"
"audiodg.exe","18812","N/A"
"svchost.exe","9260","NcdAutoSetup"
"firefox.exe","2648","N/A"
"explorer.exe","20648","N/A"
"Code.exe","24848","N/A"
"CodeHelper.exe","28060","N/A"
"conhost.exe","4216","N/A"
"Code.exe","26132","N/A"
"Code.exe","29144","N/A"
"java.exe","28668","N/A"
"conhost.exe","4716","N/A"
"conhost.exe","16888","N/A"
"powershell.exe","28216","N/A"
"Code.exe","15220","N/A"
"Code.exe","30444","N/A"
"Code.exe","29444","N/A"
"Code.exe","10400","N/A"
"Code.exe","21808","N/A"
"Code.exe","28328","N/A"
"SystemSettingsBroker.exe","27828","N/A"
"svchost.exe","28504","camsvc"
"svchost.exe","18372","RmSvc"
"SearchUI.exe","27712","N/A"
"YourPhone.exe","24112","N/A"
"svchost.exe","31308","lmhosts"
"firefox.exe","8344","N/A"
"java.exe","5520","N/A"
"conhost.exe","17784","N/A"
"firefox.exe","25856","N/A"
"firefox.exe","29044","N/A"
"firefox.exe","18688","N/A"
"WinSCP.exe","21632","N/A"
"PuTTYNG.exe","20380","N/A"
"Fork.exe","27512","N/A"
"firefox.exe","23660","N/A"
"firefox.exe","30052","N/A"
"firefox.exe","16572","N/A"
"java.exe","17728","N/A"
"conhost.exe","28640","N/A"
"firefox.exe","3056","N/A"
"svchost.exe","5936","W32Time"
"SearchProtocolHost.exe","28348","N/A"
"heidisql.exe","32884","N/A"
"heidisql.exe","19944","N/A"
"java.exe","24912","N/A"
"conhost.exe","10492","N/A"
"java.exe","14676","N/A"
"conhost.exe","31660","N/A"
"Winamax Poker.exe","16936","N/A"
"java.exe","1472","N/A"
"java.exe","27976","N/A"
"conhost.exe","22120","N/A"
"conhost.exe","17772","N/A"
"RuntimeBroker.exe","33576","N/A"
"svchost.exe","28516","ClipSVC"
"SearchFilterHost.exe","9616","N/A"
"RuntimeBroker.exe","4508","N/A"
"svchost.exe","18312","wlidsvc"
"smartscreen.exe","32384","N/A"
"cmd.exe","1608","N/A"
"conhost.exe","27064","N/A"
"tasklist.exe","23220","N/A"
```

**Comment by ayedo** [comment]
Thanks for your reply. 

I'm not sure what's going on here. I'm having trouble reproducing this problem.

The `tasklist` command on your machines does not seem to return the headers of the exported CSV even though it's configured to do so. 

Which exact version of Windows are you running?

Can you maybe check if you have any difference in output when you run the following two commands:

- `tasklist /svc /fo csv /nh` <- this one should *not* have CSV headers on the first line
- `tasklist /svc /fo csv` <- this one should have CSV headers

The headers should read `"Image Name","PID","Services"`

When you run:

` tasklist /?`

Does the parameter list contain the following two entries?

```
   /FO    format           Specifies the output format.
                           Valid values: "TABLE", "LIST", "CSV".

   /NH                     Specifies that the "Column Header" should
                           not be displayed in the output.
                           Valid only for "TABLE" and "CSV" formats
```

Anyway, I can introduce a little work around that does not depend on the headers being present, to read the data from tasklist, but I'm still very curious why this does not work on your machine.

**Comment by BenjaminBini** [comment]
So I'm using Windows 10 version 1909 (build 18363.720)

`tasklist /svc /fo csv /nh ` returns CSV without header.
`tasklist /svc /fo csv ` returns CSV with headers. It seems it did not when I ran it tjhe previous time? Or maybe I failed my copy/paste.
The first line is `"Nom de l’image","PID","Services"` because my system if in French.
Could language be the issue there? Maybe you rely on the exact headers?

When I run the help command I got those options, yes.

**Comment by BenjaminBini** [comment]
I think language is the issue indeed. 
```kotlin
 return rows.map({ row ->
        val pid = row.getValue("PID").toInt()
        val imageName = row.getValue("Image Name")
        val services = row.getValue("Services")
        val name = imageName + if (services == "N/A") "" else " ($services)"

        ProcessId(pid) to ProcessName(name)

}).toMap()
```

It looks like you get columns by there names but tasklist translate them according to user language.
Maybe you should rely on column index and not column name? Or maybe there is an option I don't know to force the command to return its result in english?

**Comment by ayedo** [comment]
Thanks! That's it. I think, you've found the bug 👍 

I don't have a lot of experience with Windows, and it never occurred to me that the output of a command line tool could change depending on the user's locale.

Stay tuned for the fix!

**Comment by ayedo** [comment]
should be fixed with release v0.2

**Comment by BenjaminBini** [comment]
It works indeed ;) Thank you!
